### ISSUE TITLE
Form inputs reset after submission in Next.js 15 when using server actions

### ISSUE THREAD

**Issue by fikkatra** [vercel/next.js#72949]
### Link to the code that reproduces this issue

https://github.com/fikkatra/next-form-poc

### To Reproduce

1. Start the app by running npm run dev
2. Open the app in the browser at http://localhost:3000
3. Fill in the form and click the submit button
4. Result: the form is emptied after the submit. All input is lost.

### Current vs. Expected behavior

I expect the form to contain the original data, since I am not resetting the form or redirecting. This is important in case of form validation errors (which I didn't include to keep the code simple).

Before submit:
<img width="239" alt="image" src="https://github.com/user-attachments/assets/9aabb58a-f2e9-4978-b826-a24bd88fc9b1">

Server console output:
```
submitForm action FormData {
  [Symbol(state)]: [
    {
      name: '$ACTION_ID_40d23cd8bd75560ac5be4fdcb74a7a5de0950cbc8e',
      value: ''
    },
    { name: 'firstName', value: 'My first name' },
    { name: 'lastName', value: 'My last name' }
  ]
}
 POST / 200 in 46ms
```

After submit:
<img width="219" alt="image" src="https://github.com/user-attachments/assets/294eeb90-cfdf-49d8-8f02-2978d6ace854">

Browser console:
`[Fast Refresh] rebuilding`

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 24.1.0: Thu Oct 10 21:03:15 PDT 2024; root:xnu-11215.41.3~2/RELEASE_ARM64_T6000
  Available memory (MB): 32768
  Available CPU cores: 8
Binaries:
  Node: 18.20.4
  npm: 10.7.0
  Yarn: 1.22.19
  pnpm: N/A
Relevant Packages:
  next: 15.0.3 // Latest available version is detected (15.0.3).
  eslint-config-next: 15.0.3
  react: 19.0.0-rc-66855b96-20241106
  react-dom: 19.0.0-rc-66855b96-20241106
  typescript: 5.6.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local), next start (local)

### Additional context

I'm not using a canary version.

**Comment by antdev1987** [comment]
but the documentation in the part of server actions and mutations says this,,    (After hydration, the browser does not refresh on form submission.)

**Comment by Marabii** [comment]
I found this blog a while ago : "https://nehalist.io/react-hook-form-with-nextjs-server-actions/" hope you find it useful @fikkatra. You may want to use the useActionState hook instead of useFormStatus they used on the blog.

**Comment by fikkatra** [comment]
Yeah, I saw that one too and followed their instructions, but then I bumped into the emptied form issue... I thought it might have something to do with a newer or more recent next or react feature, since the blog still uses useFormStatus. So that brought me here in the first place 🤷‍♀️. They are also not storing the form values. Unless I overlooked something, the blog's implementation will have the same behavior upon submitting to the server, the form will be empty.

Thanks a lot for sharing though, and for taking the time to help me out!

**Comment by thatanjan** [comment]
I am also facing this problem in the next 15. 
```json
{    "next": "15.0.2",
    "react": "19.0.0-rc-02c0e824-20241028",
    "react-dom": "19.0.0-rc-02c0e824-20241028"
}
```
But it worked fine in : 
```json
{
 "next": "14.2.13",
 "react": "^18",
 "react-dom": "^18",
}
```

**Comment by antdev1987** [comment]
this is my solution but you have to use  the "use client" 
this is the handle submit 

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>)=>{
    e.preventDefault(); // Evita el reload de la página
    const formData = new FormData(e.currentTarget);
    createEmployee(formData); // Envía los datos al server action
  }

and in the form you call your server action as normal
form action={createEmployee}
createEmployee is the server action

**Comment by fikkatra** [comment]
Since it worked in a previous version according to @thatanjan , I'm reopening this issue.

**Comment by jvondermarck** [comment]
I have the same issue with Next.JS v15... Any workaround that works ?

**Comment by jvondermarck** [comment]
> this is my solution but you have to use the "use client" this is the handle submit
> 
> ```
>   const handleSubmit = (e: React.FormEvent<HTMLFormElement>)=>{
>     e.preventDefault(); // Evita el reload de la página
>     const formData = new FormData(e.currentTarget);
>     createEmployee(formData); // Envía los datos al server action
>   }
> ```
> 
> and in the form you call your server action as normal `form action={createEmployee} onSubmit={handleSubmit}` createEmployee is the server action

This code does not work when you use zod and you want to display your input error by doing this `state?.errors?.name` for example, it does not display the errors anymore.

**Comment by fikkatra** [comment]
My current workaround (with `react-hook-forms` and `zod`) is the following, but it has some downsides (you need to handle the loading state yourself, and it prevents redirecting from the server side). Note that it does not use `useActionState` at all., as I unfortunately haven't found a way to use `useActionState` without wiping the form after submitting.

```
'use client';

import { zodResolver } from '@hookform/resolvers/zod';
import { useRouter } from 'next/navigation';
import { useState } from 'react';
import { useForm } from 'react-hook-form';
import { z } from 'zod';

import { submitForm } from '@/actions';

type MyFormData = {
  firstName: string;
  lastName: string;
};

export default function CardActivationForm() {
  const router = useRouter();

  const { handleSubmit, setError } = useForm<MyFormData>({
    resolver: zodResolver(
      z.object({
        firstName: z.string(),
        lastName: z.string(),
      }),
    ),
  });
  const [isLoading, setIsLoading] = useState(false);

  const submit = async (formData: MyFormData) => {
    setIsLoading(true);
    try {
      const state = await submitForm(formData);   // <--------- we manually call the server action here

      if (state?.status === 'success') {
        router.push('/confirmation');
      } else {
        state.errors?.forEach((error) => {
          const message = error.message;
          setError(error.path, {
            message,
          });
        });
      }
    } finally {
      setIsLoading(false);
    }
  };

  if (isLoading) {
    return <div>Loading...</div>;
  }

  return <form onSubmit={handleSubmit(submit)}>...</form>;
}
```

**Comment by thatanjan** [comment]
There might be a better solution. But here is mine: 

Custom hook to update state with `useActionState` hook
```javascript
import { useActionState, useEffect, useState } from 'react'

const useResettableActionState = ({ initialState, action }) => {
  const [state, setState] = useState(initialState)
  const [data, formAction, isPending] = useActionState(action, initialState)

  // Set the state to the data when the data is available
  useEffect(() => {
    if (!isPending) setState(data)
  }, [isPending])

  const resetState = () => setState(initialState)
  const updateState = newState => setState(prev => ({ ...prev, ...newState }))

  return {
    state,
    formAction,
    isPending,
    resetState,
    updateState,
  }
}

export default useResettableActionState
```

**action**:
```javascript
const updatePassword = async (_, formData) => {
  // Your own logic
  return {
    formFields: {} // formData in key-value pair
  }
}

```

** In the form:**

```javascript
const Page = () => {
  const { formAction, isPending, resetState, state } = useResettableActionState(action, {})
  return (
            <input
              defaultValue={state.formFields.email} // Getting the value from the state and adding add default value when user input will be clear
              type='email'
              name='email'
              className='input input-bordered w-full max-w-xs'
            />
  )
}

export default Page
```

**Comment by Marabii** [comment]
Can someone explain to me why making all inputs controlled is a bad solution ? The only input you can't make it controlled is file input, which is good enough for most cases i guess. I created a Custom controlled input wrapper and used it throughout my application.

**Comment by fikkatra** [comment]
> Can someone explain to me why making all inputs controlled is a bad solution ? The only input you can't make it controlled is file input, which is good enough for most cases i guess. I created a Custom controlled input wrapper and used it throughout my application.

I don't there's anything wrong with it: I also have a custom controlled input wrapper. I just stripped my code to the bare essentials for this GitHub issue. Unfortunately, having a custom wrapper input does not solve the issue for me. Unlike your code example, I'm using `react-hook-form` to control the input and handle validation errors.

**Comment by sm3sher** [comment]
I would recommend updating the issue title to:
### Form inputs reset after submission in Next.js 15 when using server actions
Notably, many tutorials on server actions with forms in Next.js rely on version 14, where forms retain their input values after server-side validation. In those cases, the input fields are not reset upon submission, which aligns better with user expectations during form interactions.

If this change in behavior is intentional, it should be explicitly documented as a breaking change, as it significantly impacts how developers handle forms and user input in their applications.

**Comment by VesperQuartz** [comment]
> Since it worked in a previous version according to @thatanjan , I'm reopening this issue.

Yes i can confirm it did. I came back to an old project updated to next 15 and everything just broke.

**Comment by ivi3** [comment]
It seems like Next.js never reaches a stable version, I'm always stuck with basic features!

**Comment by damien-monni** [comment]
I've stumbled across inconsistent behavior. Input resets after a form submission, but select don't.

Here is a basic reproduction: https://codesandbox.io/p/devbox/mfm7kf?file=%2Fapp%2Fpage.tsx%3A36%2C18

When you submit the form, the default value changes. On the input, you can see the value is reset, so the value is changed, but not on the select.

**Comment by Codename14** [comment]
Has anyone come up with a usable solution since then? This is really a breaking change. If anyone is using something more advanced for forms, like React Hook Form and Zod, it’s all broken. For me, the value it displays is no longer in sync. When there’s an incorrect form submission, the value disappears from the UI, but the state in Zod and React Hook Form still retains it. If it’s successful, the form is completely empty. No matter how much I revalidate the tag with a cached request, React Hook Form can’t load the default values unless I reload the page. I saw that I can return the defaults in useActionState, but why does it need to be like this? It needs to be preserved much more often than deleted; all users are using validation. How would it know off the top of its head what can be typed into an input field? Also, if the logic is more complex, like in my case, here’s my code:

action={async () => {
    const formData = getValues();
    const result = await trigger();
    if (!result) return;
    const { error } = await updateDishPrices(formData);
    if (error) {
        toast.error(error);
    } else {
        toast.success('Saved');
    }
}}

How would I write this into useActionState?

**Comment by d-reader-josip** [comment]
+1 would like this issue to be address by the Next team

**Comment by ztanner** [comment]
Hi,

As described in the React docs: https://react.dev/reference/rsc/server-functions#server-functions-with-actions

> When the Form submission succeeds, React will automatically reset the form. 

There's a discussion about this behavior in the React repo, as well as workarounds: https://github.com/facebook/react/issues/29034

Unfortunately this is outside the control of Next.js, so I'm going to be closing this issue.

**Comment by sm3sher** [comment]
It looks like this issue has been addressed in the pull request [#30728](https://github.com/facebook/react/pull/30728), which has been open for the past 5 months.
It may be worth reviewing to see if it can be merged and resolve the issue.

**Comment by devdatkumar** [comment]
**solution 1:** Return `errors `and `rawData` from the server action and repopulate the form using HTML `defaultValue`
```tsx     
errors: validatedFields.error.flatten().fieldErrors,
rawData: formData
 ```
However, this approach triggers unnecessary API requests for every invalid input. For example, if you have 20 fields and make 5 incorrect submissions before a successful one, that results in 6 API requests instead of just 1.

To avoid this, I considered validating data types on the front end.

**Solution 2:** not sure,  if this is efficient or not. 
```tsx
type FieldError = {
  email?: string[] | undefined;
  password?: string[] | undefined;
};

export default function SigninForm() {
  const [user, setUser] = useState({ email: "", password: "" });
  const [error, setError] = useState<FieldError | undefined>({});

  const [state, dispatch, isPending] = useActionState(signinAction, undefined);

  const handleAction = (formData: FormData) => {

    const validationResult = signinSchema.safeParse(
      Object.fromEntries(formData)
    );

    if (!validationResult.success) {

      setError(validationResult.error.flatten().fieldErrors);

      setUser({
        email: formData.get("email")?.toString() ?? "",
        password: formData.get("password")?.toString() ?? "",
      });

      return;
    }

    setError({});
    dispatch(formData);
    setUser({ email: "", password: "" });
  };
```

**Comment by lpkobamn** [comment]
i feel like i live in a simulation.

**Comment by budokans** [comment]
The only workaround I've found so far that isn't _totally_ disgusting is to add a `key` prop that will change after each submission to the input element's parent element. I'm using a `Fragment` specifically to wrap these, partly because the input elements often have siblings, and also so I can more easily get rid of them all if a better solution reveals itself. A contrived example:

```
interface CreateUserRawFormData {
  readonly username: FormDataEntryValue | null
}

<Fragment
  key={formData.firstName}
>
  <input
    defaultValue={typeof formData.username === "string" ? formData.username : undefined}
  />
</Fragment>
```

I realise that this is an anti-pattern and I loathe it just as much as you do.

**Comment by urboifox** [comment]
The default behavior in server actions in general is resetting the forms after submission.
But i was using Next.js since version 13, and it always kept the original values after form actions. and people started asking questions on "how to reset nextjs form after server action" (you can google that, you will get a lot of results).

however, in version 15, this behavior changed, which is so confusing because i couldn't it mentioned in the release notes.

was it an issue in older versions? or a feature in version 15? idk.

anyways, i found a solution (i think on reddit) where you replace the `action` attribute with the `onSubmit`, then calling `onTransition` and pass it your action.

this code smells, but it works as expected.
heres the snippet:

```tsx
<form
    onSubmit={(e) => {
        e.preventDefault();
        startTransition(() => myAction(new FormData(e.currentTarget)));
    }}
    // action={myAction}
>
```

btw this ruined my attempt to migrate to next 15, because im too lazy to change all the `action` attributes on my forms.

In SvelteKit however, you can handle that with an `onreset` attribute:

```svelte
<form
    method="post"
    action="?/login"
    onreset={(e) => e.preventDefault()}
>
```

I tried it in nextjs, but it didnt work. i think they are not triggering the builtin `onreset` when they reset the form like sveltekit does. however it would be nice if something like this exists, then you can decide whether you want your form to reset or no.

**Comment by urboifox** [comment]
The default behavior in server actions in general is resetting the forms after submission.
But in nextjs since version 13, forms kept the original values after form actions.
however, in version 15, this behavior changed, which is so confusing because I found no one talking about it.

I think this is a react related thing, not nextjs though.

Anyways, there are a couple of solutions.
You can use the `onSubmit` attribute instead of `action`, then call `onTransition` and pass it your action function.

```tsx
<form
    onSubmit={(e) => {
        e.preventDefault();
        startTransition(() => myAction(new FormData(e.currentTarget)));
    }}
    // action={myAction}
>
```

Another solution is to return the form data from your server action, and set it again in your form as default values.

```ts
export async function myAction(prevState: any, formData: FormData) {
    // do your stuff here
    return {
        formData,
        success: true,
        data: // the actual data you want to return
    };
}
```

```tsx
const [state, action, pending] = useActionState(myAction, { success: false })
...
<input name="email" defaultValue={state.formData?.get("email")} />
```

In SvelteKit however, you can handle that with an `onreset` attribute:

```svelte
<form
    method="post"
    action="?/login"
    onreset={(e) => e.preventDefault()}
>
```

I tried it in react, but it didnt work. It would be nice if something like this works in react, then you can decide whether you want your form to reset or no.

**Comment by devdatkumar** [comment]
> In SvelteKit however, you can handle that with an `onreset` attribute:
> 
```html
 <form
     method="post"
     action="?/login"
     onreset={(e) => e.preventDefault()}
 >
```
> I tried it in react, but it didnt work. It would be nice if something like this works in react, then you can decide whether you want your form to reset or no.


**We could write a wrapper around the form element and mimic its behavior!**

**Comment by desjardinsalec** [comment]
> I don't like the second solution so much, because it doesn't work well with image and select inputs. and if you have a kinda complex form, goodluck with that.

Why does it not work with selects? I am running that issue as well.

**Comment by robin-a-meade** [comment]
@desjardinsalec Setting the `key` attribute in addition to `defaultValue` makes it work with `select`. See https://github.com/facebook/react/issues/30580 . Note, however, that the `key` attribute is [not documented](https://react.dev/reference/react-dom/components/select#select).

**Comment by desjardinsalec** [comment]
@robin-a-meade I'll have to test that. I have been battling this issue since last week. Thanks for the suggestion!

**Comment by urboifox** [comment]
> > I don't like the second solution so much, because it doesn't work well with image and select inputs. and if you have a kinda complex form, goodluck with that.
> 
> Why does it not work with selects? I am running that issue as well.

`<select>` elements do not automatically reapply defaultValue after a rerender. you need to add a key prop to force a rerender

for example
```tsx
<select 
  defaultValue={state.formData?.get("degree") as string}
  key={state.formData?.get("degree") as string}
>
```

**Comment by desjardinsalec** [comment]
@urboifox thanks! I was able to get it working with `key` yesterday based on the linked issue. I've never seen key on a select...

**Comment by JClackett** [comment]
What do we do for controlled components?

**Comment by urboifox** [comment]
@JClackett Nothing, they are out of the scope of this problem. Because they already persist their values even if the form resets.
and if you wanna reset them, you have to do it manually by removing the values from your state.

**Comment by JClackett** [comment]
@urboifox thats not true, if I have a controlled select component inside of a form that uses action even if I have value set, it resets..

**Comment by urboifox** [comment]
@JClackett I just tried it in a demo page, and you are right.
i solved it by re-rendering the form after submission tho, maybe there is a better solution, but that's what i thought of for now.

```tsx
"use client";
import { useState } from "react";
import { action } from "./action";

export default function Form() {
    const [selected, setSelected] = useState("0");
    const [formKey, setFormKey] = useState(0);

    return (
        <div>
            <form
                key={formKey}
                action={(formData) => {
                    action(formData);
                    setFormKey(formKey + 1);
                }}
            >
                <select
                    value={selected}
                    onChange={(e) => setSelected(e.target.value)}
                    name="select"
                >
                    <option value="0">Select</option>
                    <option value="1">Option 1</option>
                    <option value="2">Option 2</option>
                    <option value="3">Option 3</option>
                </select>
                <button>Submit</button>
            </form>
        </div>
    );
}
```

everything worked as expected in this demo, we can also limit the key to only the select elements to prevent unnecessary re-renders.
